plasmapy/plasma/grids.py

```diff
@@ -0,0 +1,330 @@
+            self._make_grid(start=start, stop=stop, num=num)
+        else:
+            self._load_grid(grid=grid, units=units)
+
+        # Setup method contains other initialization stuff
+        self._setup()
+
+    def _setup(self):
+        # Properties
+        self._coordinate_system = None
+        self._regular_grid = None
+
+    @property
+    def grid(self):
+        """Grid of positional values."""
+        return self._grid
+
+    @property
+    def regular_grid(self):
+        """
+        Value of regular_grid
+        If None, calculate
+        """
+        if self._regular_grid is None:
+            self._detect_regular_grid()
+        return self._regular_grid
+
+    @property
+    def shape(self):
+        return self._grid.shape
+
+    @property
+    def unit(self):
+        if self._units[0] == self._units[1] and self._units[0] == self._units[2]:
+            return self._units[0]
+        else:
+            raise ValueError("Array dimensions do not all have the same "
+                             f"units: {self._units}")
+
+    @property
+    def units(self):
+        return self._units
+
+    @property
+    def unit0(self):
+        return self._units[0]
+
+    @property
+    def unit1(self):
+        return self._units[1]
+
+    @property
+    def unit2(self):
+        return self._units[2]
+
+    @property
+    def ax0(self):
+        return self._grid[:, 0, 0, 0]
+
+    @property
+    def ax1(self):
+        return self._grid[0, :, 0, 1]
+
+    @property
+    def ax2(self):
+        return self._grid[0, 0, :, 2]
+
+    @property
+    def dax0(self):
+        return np.mean(np.gradient(self.ax0))
+
+    @property
+    def dax1(self):
+        return np.mean(np.gradient(self.ax1))
+
+    @property
+    def dax2(self):
+        return np.mean(np.gradient(self.ax2))
+
+    @property
+    def arr0(self):
+        return self._grid[..., 0]
+
+    @property
+    def arr1(self):
+        return self._grid[..., 1]
+
+    @property
+    def arr2(self):
+        return self._grid[..., 2]
+
+    def _validate(self):
+        """
+        Checks to make sure that the grid parameters are
+        consistent with the coordinate system and units selected
+
+        """
+        return True
+
+    def _load_grid(self, grid=None, units=None):
+
+        # If single values are given, expand to a list of appropriate length
+        if isinstance(units, u.core.Unit):
+            units = [units] * 3
+
+        # Determine units from grid or units keyword (in that order)
+        if hasattr(grid, 'unit'):
+            self._units = [grid.unit] * 3
+        elif units is not None:
+            self._units = units
+        else:
+            self._units = [u.dimensionless_unscaled] * 3
+
+        self._grid = grid
+
+        # Check to make sure that the object created satisfies any
+        # requirements: eg. units correspond to the coordinate system
+        self._validate()
+
+    def _make_grid(self, start=0, stop=1, num=100):
```

Comment:
Some comment as I made for `__init__`, `start` and `stop` should be required arguments.